workflows learning path (#22263)
cloudflare/cloudflare-docs: src/content/docs/learning-paths/workflows-course/series/workflows-3.mdx

        showMoreVideos={false}
        chapters={{
          "Intro": "0s",
          "Interaction code": "1m53s",
          "Leaderboard code": "3m21s",
          "Wrangler.toml": "3m41s",
          "Puntificator: Using AI to develop more puns automatically": "4m46s",
          "Nested Workflows": "5m55s",
          "Npx wrangler workflow instances": "6m37s",
          "Conclusion": "7m41s"
        }}
      />

      **Related content**

      If you want to dive into detail, refer to the following pages:

    - [Source code for the Punderful repository](https://github.com/craigsdennis/punderful-workflows)
    - [Cloudflare Workflows](/workflows/)
    - [Cloudflare Workers AI](/workers-ai/)

    </Card>
  </TabItem>
  
  <TabItem label="Step-by-step tutorial">

      Workflows allow you to kick off asynchronous processes without blocking the user. This is demonstrated in the `addInteraction` function, which creates a new instance of the `INTERACTION` workflow.

      Locate the `addInteraction` function in `src/index.tsx`:
      
      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/index.tsx#L237)

>       This function is called when a user interacts with a pun (e.g., likes it). Instead of performing the interaction logic directly, it offloads the work to a workflow.

Reviewer: ```suggestion
      This function is called when a user interacts with a pun (for example, likes it). Instead of performing the interaction logic directly, it offloads the work to a workflow.
```

      ### Wrangler.toml Configuration

      The `wrangler.toml` file is used to configure your Worker and Workflows. This includes defining bindings to resources like KV namespaces and setting up triggers for workflows.

      Open `wrangler.toml` and find the `[triggers]` section.
      
      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/wrangler.toml#L68)

      The `crons` array allows you to define scheduled triggers for your main Worker. The example shows a cron job configured to run every 30 minutes.

      Locate the `scheduled` handler in your main Worker code (`src/index.tsx`). This handler is executed when a cron trigger fires.

      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/index.tsx#L315)

      This handler creates an instance of the `LEADERBOARD_WORKFLOW`, initiating the leaderboard update process on a schedule.

      ### Puntificator: Using AI to Develop More Puns Automatically

      Workflows can also be used for more complex, multi-step processes, including interacting with AI models. The `PuntificatorWorkflow` is an example that leverages AI to generate and evaluate new puns.

>       Examine the `PuntificatorWorkflow` definition in `src/workflows/puntificator.ts`.

Reviewer: ```suggestion
      Examine the `PuntificatorWorkflow` definition in `src/workflows/puntificator.ts`.
      
```

      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/workflows/puntificator.ts)

      This workflow includes steps to:

      1.  Retrieve trending puns.
      2.  Create a new pun based on trends using an AI model.
      3.  Judge the quality of the created pun using another AI model.
      4.  Save the pun if it meets a certain rating threshold.

      Crucially, this workflow includes a `step.sleep` call:

      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/workflows/puntificator.ts#L135)

>       This step pauses the workflow execution for a specified duration. This is useful for waiting to see user feedback on a published pun before potentially taking further action based on its popularity.

Reviewer: ```suggestion
      This step pauses the workflow execution for a specified duration. This is useful for waiting to consider user feedback on a published pun before potentially taking further action based on its popularity.
```

      ### Nested Workflows

      Workflows can initiate other workflows, allowing you to compose complex processes from smaller, modular units.

      In the `PuntificatorWorkflow`, find where it calls the `PUBLISH` workflow.
      
      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/workflows/puntificator.ts#L115)

      This demonstrates how one workflow can trigger another, enabling the separation of concerns and modular design.

      Examine the `PublishWorkflow` in `src/workflows/publish.ts`.

      [See here](https://github.com/craigsdennis/punderful-workflows/blob/main/src/workflows/publish.ts)

      This workflow handles the logic for publishing a pun, likely involving saving it to the database and making it visible on the site.

      ### Workflow Instances

      You can trigger workflows manually and inspect their execution status and output using the `wrangler` command-line tool.

      To trigger the `PuntificatorWorkflow` manually:

      ```bash
      npx wrangler workflows trigger puntificator
      ```

>       This command will queue an instance of the workflow. You will see a success message and the instance ID.

Reviewer: ```suggestion
      This command will queue an instance of the workflow. You will receive a success message and the instance ID.
```
